src/server.c

```diff
@@ -602,31 +603,99 @@ void nvnc_send_cut_text(struct nvnc* server, const char* text, uint32_t len)
 static int on_client_cut_text(struct nvnc_client* client)
 {
 	struct nvnc* server = client->server;
+	nvnc_cut_text_fn fn = server->cut_text_fn;
 	struct rfb_cut_text_msg* msg =
 	        (struct rfb_cut_text_msg*)(client->msg_buffer +
 	                                   client->buffer_index);
 
-	if (client->buffer_len - client->buffer_index < sizeof(*msg))
+	size_t left_to_process = client->buffer_len - client->buffer_index;
+
+	if (left_to_process < sizeof(*msg))
 		return 0;
 
 	uint32_t length = ntohl(msg->length);
+	uint32_t max_length = MAX_CUT_TEXT_SIZE;
+
+	assert(length != 0);
 
 	/* Messages greater than this size are unsupported */
-	if (length > MSG_BUFFER_SIZE - sizeof(*msg)) {
+	if (length > max_length) {
+		log_error("Copied text length (%d) is greater than max supported length (%d)\n",
+			length, max_length);
 		stream_close(client->net_stream);
 		client_unref(client);
 		return 0;
 	}
 
-	if (client->buffer_len - client->buffer_index < sizeof(*msg) + length)
-		return 0;
+	size_t msg_size = sizeof(*msg) + length;
+
+	if (msg_size <= left_to_process) {
+		if (fn)
+			fn(server, msg->text, length);
+
+		return msg_size;
+	}
+
+	if (client->cut_text.buffer == NULL) {
+		client->cut_text.buffer = malloc(length);
+		if (client->cut_text.buffer == NULL) {
+			log_error("OOM: %m\n");
+			stream_close(client->net_stream);
+			client_unref(client);
+			return 0;
+		}
 
+		size_t partial_size = left_to_process - sizeof(*msg);
+
+		memcpy(client->cut_text.buffer, msg->text, partial_size);
+
+		client->cut_text.length = length;
+		client->cut_text.index = partial_size;
+		return left_to_process;
+	}
+
+	/* Control should never reach this */
```

Comment:
If that's true then the `if` in line 639 is useless and should be replaced with `assert(client->cut_text.buffer == NULL)`.

Reply:
> ... should be replaced with `assert(client->cut_text.buffer == NULL)`.

this^^.